core/banking/gov_transfers.go

```diff
@@ -0,0 +1,271 @@
+	timepoints := []int64{}
+	now := e.timeService.GetTimeNow()
+	for k := range e.scheduledGovernanceTransfers {
+		if now.UnixNano() >= k {
+			timepoints = append(timepoints, k)
+		}
+	}
+
+	for _, t := range timepoints {
+		transfers := e.scheduledGovernanceTransfers[t]
+		for _, gTransfer := range transfers {
+			amt, err := e.processGovernanceTransfer(ctx, gTransfer)
+			if err != nil {
+				gTransfer.Status = types.TransferStatusStopped
+				e.broker.Send(events.NewGovTransferFundsEventWithReason(ctx, gTransfer, amt, err.Error()))
+			}
+		}
+		delete(e.scheduledGovernanceTransfers, t)
+	}
+}
+
+func (e *Engine) distributeRecurringGovernanceTransfers(ctx context.Context, newEpoch uint64) {
+	var (
+		transfersDone = []events.Event{}
+		doneIDs       = []string{}
+	)
+
+	for _, gTransfer := range e.recurringGovernanceTransfers {
+		if gTransfer.Config.RecurringTransferConfig.StartEpoch > newEpoch {
+			continue
+		}
+
+		amount, err := e.processGovernanceTransfer(ctx, gTransfer)
+		if err != nil {
+			e.log.Error("error calculating transfer amount for governance transfer", logging.Error(err))
+			gTransfer.Status = types.TransferStatusStopped
+			transfersDone = append(transfersDone, events.NewGovTransferFundsEventWithReason(ctx, gTransfer, amount, err.Error()))
+			doneIDs = append(doneIDs, gTransfer.ID)
+			continue
+		}
+
+		if gTransfer.Config.RecurringTransferConfig.EndEpoch != nil && *gTransfer.Config.RecurringTransferConfig.EndEpoch == e.currentEpoch {
+			gTransfer.Status = types.TransferStatusDone
+			transfersDone = append(transfersDone, events.NewGovTransferFundsEvent(ctx, gTransfer, amount))
+			doneIDs = append(doneIDs, gTransfer.ID)
+			continue
+		}
+		e.broker.Send(events.NewGovTransferFundsEvent(ctx, gTransfer, amount))
+	}
+
+	if len(transfersDone) > 0 {
+		for _, id := range doneIDs {
+			e.deleteGovTransfer(id)
+		}
+		e.broker.SendBatch(transfersDone)
+	}
+}
+
+func (e *Engine) deleteGovTransfer(ID string) {
+	index := -1
+	for i, rt := range e.recurringGovernanceTransfers {
+		if rt.ID == ID {
+			index = i
+			break
+		}
+	}
+	if index >= 0 {
+		e.recurringGovernanceTransfers = append(e.recurringGovernanceTransfers[:index], e.recurringGovernanceTransfers[index+1:]...)
+		delete(e.recurringGovernanceTransfersMap, ID)
+	}
+}
+
+func (e *Engine) NewGovernanceTransfer(ctx context.Context, ID, reference string, config *types.NewTransferConfiguration) error {
+	var err error
+	var amount *num.Uint
+	var gTransfer *types.GovernanceTransfer
+
+	defer func() {
+		if err != nil {
+			e.broker.Send(events.NewGovTransferFundsEventWithReason(ctx, gTransfer, amount, err.Error()))
+		} else {
+			e.broker.Send(events.NewGovTransferFundsEvent(ctx, gTransfer, amount))
+		}
+	}()
+	now := e.timeService.GetTimeNow()
+	gTransfer = &types.GovernanceTransfer{
+		ID:        ID,
+		Reference: reference,
+		Config:    config,
+		Status:    types.TransferStatsUnspecified,
```

Comment:
Should this be `TransferStatusPending` so that the `recurring` path has it set to something?